### PR TITLE
docs: update custom bentoserver guide 

### PR DESCRIPTION
**Pull request by parano**
## What does this PR address?

* Added section about usage of the `route` parameter in `api` decorated
* updated example code for mount_asgi/wsgi_apps with the latest examples here: https://github.com/bentoml/gallery/pull/148

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
  those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

## Who can help review?

Feel free to tag members/contributors who can help review your PR.

### CODE REVIEW

File: docs/source/guides/server.rst [hunk 1]
```diff
@@ -2,23 +2,68 @@
...
+By default, the inference APIs are generated from the ``@api`` defined within a
+``bentoml.Service``. The URL route for the inference API is determined by the function
+name. Take the sample service from our tutorial for example, the function name ``classify``
+will be used for generating the REST API URL ``/classify``:
+
+.. code-block:: python
+
+    svc = bentoml.Service("iris_classifier", runners=[iris_clf_runner])
+
+    @svc.api(input=NumpyNdarray(), output=NumpyNdarray())\
```
Comment (parano) [inline docs/source/guides/server.rst:27]:
```suggestion
    @svc.api(input=NumpyNdarray(), output=NumpyNdarray())
```

File: docs/source/guides/server.rst [hunk 2]
```diff
@@ -2,23 +2,68 @@
...
+    def classify(input_arr):
+        ...
+
+However, user can customize this URL endpoint via the ``route`` option in the
+``bentoml.Service#api`` decorator. For example, the following code will generate an
+endpoint with URL `/v1/models/iris_classifier`, regardless of the API function name:
```
Comment (parano) [inline docs/source/guides/server.rst:33]:
```suggestion
endpoint with URL ``/v1/models/iris_classifier``, regardless of the API function name:
```

File: docs/source/guides/server.rst [hunk 3]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
...
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    # For demo purpose, here's an identical inference endpoint implemented via FastAPI
+    @fastapi_app.post("/predict_fastapi")
+    def predict(features: IrisFeatures):
+        input_df = pd.DataFrame([features.dict()])
+        results = iris_clf_runner.predict.run(input_df)
+        return { "prediction": results.tolist()[0] }
 
-Here’s an example of mounting a FastAPI app alongside BentoML
+    # For demo purpose, here's an identical inference endpoint implemented via FastAPI
```
Comment (parano) [inline docs/source/guides/server.rst:144]:
```suggestion
    # BentoML Runner's async API is recommended for async endpoints
```

File: docs/source/guides/server.rst [hunk 4]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
...
+    async def predict_async(features: IrisFeatures):
+        input_df = pd.DataFrame([features.dict()])
+        results = await iris_clf_runner.predict.async_run(input_df)
+        return { "prediction": results.tolist()[0] }
 
-.. code-block:: python
 
-    import bentoml
-    from bentoml.io import JSON
+This goes the same for ASGI web applications built with any ASGI frameworks other than
```
Comment (parano) [inline docs/source/guides/server.rst:152]:
```suggestion
This goes the same for ASGI web applications built with any ASGI framework other than
```

File: docs/source/guides/server.rst [hunk 5]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
     svc.add_asgi_middleware(HTTPSRedirectMiddleware)
 
 
-Customize API Server
---------------------
+Fully Customized Endpoints
+--------------------------
 
-BentoML provides first-class support for bundling existing WSGI or ASGI app into a BentoServer.
+BentoML provides first-class support for bundling existing WSGI or ASGI app into a
```
Comment (ssheng) [inline docs/source/guides/server.rst:88]:
```suggestion
BentoML provides first-class support for mounting existing WSGI or ASGI apps onto the
```

File: docs/source/guides/server.rst [hunk 6]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
...
+BentoServer's web serving layer, to fully extend BentoML's REST API to adapt to your
+needs.
```
Comment (ssheng) [inline docs/source/guides/server.rst:90]:
```suggestion
web serving layer, to enable common use cases such as serving existing Python web applications alongside
the models, performing custom authentication and authorization, handling GET requests and web UIs, or
providing streaming capabilities.
```

File: docs/source/guides/server.rst [hunk 7]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
...
 
-Bundle WSGI app (e.g. Flask)
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Use this method if you want to serve existing Python Web applications with your models,
+or if you need to customize the API request handling to handle authentication, GET requests,
+custom web UI, or provide HTTP streaming endpoints.
```
Comment (ssheng) [inline docs/source/guides/server.rst:94]:
```suggestion
```

File: docs/source/guides/server.rst [hunk 8]
```diff
@@ -2,23 +2,68 @@
 Customize BentoServer
 =====================
 
+BentoServer architecture runs the BentoML Service API in an `ASGI <https://asgi.readthedocs.io/en/latest/>`_
+web serving layer and Runners in a separate work process pool manged by BentoML. The ASGI web
```
Comment (ssheng) [inline docs/source/guides/server.rst:6]: For the recurring term of "web serving layer", should we just call it "web server" or "http server"?

File: docs/source/guides/server.rst [hunk 9]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
...
+        input_df = pd.DataFrame([features.dict()])
+        results = await iris_clf_runner.predict.async_run(input_df)
+        return { "prediction": results.tolist()[0] }
 
-.. code-block:: python
 
-    import bentoml
-    from bentoml.io import JSON
+This goes the same for ASGI web applications built with any ASGI framework other than
+FastAPI as well.
```
Comment (ssheng) [inline docs/source/guides/server.rst:153]:
```suggestion
In addition to FastAPI, application mounting is supported for any ASGI web applications built with any frameworks adhering to the ASGI standards.
```

File: docs/source/guides/server.rst [hunk 10]
```diff
@@ -37,63 +82,115 @@ For example, you can add do:
...
 
 
-Customize API Server
---------------------
+Fully Customized Endpoints
+--------------------------
 
-BentoML provides first-class support for bundling existing WSGI or ASGI app into a BentoServer.
+BentoML provides first-class support for bundling existing WSGI or ASGI app into a
+BentoServer's web serving layer, to fully extend BentoML's REST API to adapt to your
```
Comment (parano) [inline docs/source/guides/server.rst:89]:
```suggestion
BentoML Server's web serving layer, to fully extend BentoML's REST API to adapt to your
```

File: docs/source/guides/server.rst [hunk 11]
```diff
@@ -2,23 +2,68 @@
 Customize BentoServer
 =====================
 
+BentoServer architecture runs the BentoML Service API in an `ASGI <https://asgi.readthedocs.io/en/latest/>`_
+web serving layer and Runners in a separate worker process pool managed by BentoML. The ASGI web
```
Comment (parano) [inline docs/source/guides/server.rst:6]:
```suggestion
BentoML Server runs the Service API in an `ASGI <https://asgi.readthedocs.io/en/latest/>`_
web serving layer and puts Runners in a separate worker process pool managed by BentoML. The ASGI web
```

File: docs/source/guides/server.rst [hunk 12]
```diff
@@ -2,23 +2,66 @@
...
-BentoServer endpoints.
+Since the web serving layer is built with the Python ASGI protocol, users can use the
+:code:`bentoml.Service#add_asgi_middleware` API to mount arbitrary
+`ASGI middleware <https://asgi.readthedocs.io/en/latest/specs/main.html>`_ to change
+anything they may need to customize in the HTTP request to response lifecycle, such as
+manipulating the request headers, modifying the response status code, or authorizing access to an endpoint.
 
-Users can implement their own ASGI middleware class, but before you do so, make sure
+Users can implement their own ASGI middleware class and use with BentoML,
 to checkout the existing middleware built by the Python community, such as:
```
Comment (parano) [inline docs/source/guides/server.rst:65]:
```suggestion
Users can not only implement their own ASGI middleware class,
but also use existing middleware built by the Python web development community, such as:
```

**Review by ssheng — APPROVED** [review]

File: docs/source/guides/server.rst (hunk 8 above)
Comment (ssheng) [reply docs/source/guides/server.rst]: We can discuss separately. No reason to hold on to this PR.

**Pull request merged**